server/neptune/workflows/internal/deploy/revision/queue/pull_rebaser.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/pkg/errors"
+	key "github.com/runatlantis/atlantis/server/neptune/context"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities/github"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities/terraform"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config/logger"
+	"go.temporal.io/sdk/workflow"
+)
+
+// GH API Ratelimit resets every hour, 2 hours should be enough for the GH API call failures to autoresolve
+const RebaseGithubScheduleToCloseTimeout = 2 * time.Hour
+
+type githubRebaseActivities interface {
+	GithubListOpenPRs(ctx context.Context, request activities.ListOpenPRsRequest) (activities.ListOpenPRsResponse, error)
+	GithubListModifiedFiles(ctx context.Context, request activities.ListModifiedFilesRequest) (activities.ListModifiedFilesResponse, error)
+}
+
+type prRevisionSetterActivities interface {
+	SetPRRevision(ctx context.Context, request activities.SetPRRevisionRequest) (activities.SetPRRevisionResponse, error)
+}
+
+type PullRebaser struct {
+	GithubActivities           githubRebaseActivities
+	PrRevisionSetterActivities prRevisionSetterActivities
+}
+
+func (p *PullRebaser) RebaseOpenPRsForRoot(ctx workflow.Context, repo github.Repo, root terraform.Root) error {
+	originalOpts := workflow.GetActivityOptions(ctx)
+
+	// Configure github specific ScheduleToClose timeout for GH API call failures to autoresolve
+	// TODO: Add alarms on GithubListOpenPRs and GithubListModifiedFiles API call failures
+	ghOpts := originalOpts
+	ghOpts.ScheduleToCloseTimeout = RebaseGithubScheduleToCloseTimeout
+	ctx = workflow.WithActivityOptions(ctx, ghOpts)
+
+	// list open PRs
+	var listOpenPRsResp activities.ListOpenPRsResponse
+	err := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListOpenPRs, activities.ListOpenPRsRequest{
+		Repo: repo,
+	}).Get(ctx, &listOpenPRsResp)
+	if err != nil {
+		return errors.Wrap(err, "listing open PRs")
+	}
+
+	// spawn activities to list modified fles in each open PR in parallel
+	futureByPullNum := map[github.PullRequest]workflow.Future{}
+	for _, pullRequest := range listOpenPRsResp.PullRequests {
+		future := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListModifiedFiles, activities.ListModifiedFilesRequest{
+			Repo:        repo,
+			PullRequest: pullRequest,
+		})
+		futureByPullNum[pullRequest] = future
+	}
+
+	// resolve the futures and check if an open PR needs to be rebased
+	prsToRebase := []github.PullRequest{}
+	for pullRequest, future := range futureByPullNum {
+		var result activities.ListModifiedFilesResponse
+
+		// list modified files should not fail unless we hit the ratelimit which should autoresolve once our ratelimit revives in an hour max
+		// If it errors out due to any other reason, let's be preventive and rebase this PR as well
+		listFilesErr := future.Get(ctx, &result)
+		if listFilesErr != nil {
```

Comment:
This comment says we should be preventative and rebase in the case of non-rate limit failures, however, I don't see any code differentiating between the errors.  Also, retries are infinite so we'd never really error out? 

Let's pick a direction and commit.

Reply:
Since we're using a custom tq configured to work within our API budget for the rebase flow, we can leave the retry policy as is similar to our other GH API activities in the deploy workflow imo